### PR TITLE
ghc.withPackages: install documentation to -with-packages output

### LINKED ISSUE #76837: NIX_GHC_DOCDIR not set properly
**Describe the bug**
When creating a Haskell development environment like this:
```
# shell.nix
{ pkgs ? import <nixpkgs> { } }:

let
  all-hies = import (fetchTarball "https://github.com/infinisil/all-hies/tarball/master") {};
  hie = all-hies.selection { selector = p: { inherit (p) ghc865; }; };

in
  pkgs.haskellPackages.shellFor {
    buildInputs = with pkgs; [
      hie
      cabal-install
    ];

    packages = (hp: with hp; [ gloss ]);
    withHoogle = true;
  }
```

with `withHoogle = true`, the doc is generated in `${ghcEnv}/share/doc/hoogle/html`, however the `NIX_GHC_DOCDIR` is set to `${ghcEnv}/share/doc/ghc/html`.
The former does not even exist (no such file or directory error on `ls $NIX_GHC_DOCDIR`).

See this line:

https://github.com/NixOS/nixpkgs/blob/ff882242ea70502696b551c7c9bfb933a250c60f/pkgs/development/haskell-modules/make-package-set.nix#L322

Note that I have seen https://github.com/NixOS/nixpkgs/issues/20730 and think that it is a different issue.

This issue prevents from using HIE in a nix-shell. See https://discourse.nixos.org/t/cant-make-hie-work-on-vs-code-with-nix/5277/

**To Reproduce**
Use previous `shell.nix`
```
nix-shell --run "ls $NIX_GHC_DOCDIR"
nix-shell --run "whateverfolderghcisin/share/doc/hoogle"
```

**Expected behavior**
`nix-shell --run "ls $NIX_GHC_DOCDIR"` should point to `${ghcEnv}/share/doc/hoogle/html`

**Metadata**
```
% nix run nixpkgs.nix-info -c nix-info -m                                                                                                                                                                                                  ~
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.91, NixOS, 19.09.1686.69ed29f5f41 (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3`
 - channels(arsleust): `"home-manager"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### LINKED ISSUE #150666: haskellPackages.shellFor creates invalid NIX_GHC_DOCDIR
### Describe the bug

haskellPackages.shellFor function creates environment with `NIX_GHC_DOCDIR`
variable set to non-existent path.

### Steps To Reproduce

````
$ cat bad.nix
let
  pkgs = import ./. {};
in pkgs.haskellPackages.shellFor {
  packages = p: [ p.lens ];
}
$ nix-shell bad.nix --run 'ls -al $NIX_GHC_DOCDIR'
ls: cannot access '/nix/store/5hjpspgqm47hqwdjrryrqglp0zfz0b0a-ghc-8.10.7-with-packages/share/doc/ghc/html': No such file or directory
````

### Expected behavior

`$NIX_GHC_DOCDIR` points to directory with html documentation of all
dependencies.

### Notify maintainers

@peti
@cdepillabout
@expipiplus1
@maralorn
@sternenseemann
@cstrahan

### Metadata

````
$ git show
commit 290ab1ac45f92825613c5f24cc4896dd6524b748
Merge: d77cd852ab1 e06f742bebd
Author: Maciej Krüger <mkg20001@gmail.com>
Date:   Sun Dec 12 08:56:08 2021 +0100

    Merge pull request #150379 from r-ryantm/auto-update/xviewer
````

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.7`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.5pre20211007_844dd90`
 - channels(kaction): `"nixpkgs-21.11pre332093.7fad01d9d5a"`
 - nixpkgs: `/home/kaction/.nix-defexpr/channels/nixpkgs`
```

### PR DESCRIPTION
**Pull request by sternenseemann**
* Will make it so that GHC.Paths's docdir NIX_GHC_DOCDIR points to an actual directory.

* Documentation of all packages in the environment is available in `$out/share/doc`.

This has previously been attempted in #76842 and reverted in #77442, since documentation can collide when the libraries wouldn't (thanks to the hash in the lib filename). `symlinkJoin` allows collision, so this solution should be akin to #77523 (minus `buildEnv`, one step at a time). As an escape hatch, the old behavior can be restored using `installDocumentation = false`.

Collision in the documentation only happen if the dependency closure of the given packages has more than one different derivation for the same library of the very same version. I'm personally inclined not to claim that our infrastructure does anything sensible in this case. Additionally, the documentation is likely largely the same in such cases (unless it is heavily patched).

Resolves #150666.
Resolves #76837.
Closes #150968.
Closes #77523.

cc @matthewbauer 

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/haskell-modules/with-packages-wrapper.nix [hunk 1]
```diff
@@ -4,6 +4,7 @@
 # platform that has NCG support
 , useLLVM ? false
 , withHoogle ? false
+, installDocumentation ? true
```
Comment (cdepillabout) [inline pkgs/development/haskell-modules/with-packages-wrapper.nix:7]: Could you add a comment on this option?  Maybe just the same text from the release-notes file?

**Review by maralorn — APPROVED** [review]
LGTM, just wondering, do you have a concrete use case in mind for this?
Or is it just that people can direct their browsers to that directory to find all docs in one place?

File: pkgs/development/haskell-modules/with-packages-wrapper.nix (hunk 1 above)
Comment (sternenseemann) [reply pkgs/development/haskell-modules/with-packages-wrapper.nix:7]: Done.

**Comment by sternenseemann** [comment]
@maralorn for me it is mostly about correctness, i.e. having it look as normal as possible. Also it has been requested a couple of times now and there really is no cost, as we have the doc outputs in the dependency closure already anyways.

**Comment by maralorn** [comment]
Thx!

**Pull request merged**